sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -158,25 +158,34 @@ case class AdaptiveSparkPlanExec(
       var result = createQueryStages(currentPhysicalPlan)
       val events = new LinkedBlockingQueue[StageMaterializationEvent]()
       val errors = new mutable.ArrayBuffer[SparkException]()
+      val runningStages = util.Collections.synchronizedSet(new util.HashSet[QueryStageExec]())
       var stagesToReplace = Seq.empty[QueryStageExec]
       while (!result.allChildStagesMaterialized) {
         currentPhysicalPlan = result.newPlan
         if (result.newStages.nonEmpty) {
           stagesToReplace = result.newStages ++ stagesToReplace
           executionId.foreach(onUpdatePlan)
 
-          // Start materialization of all new stages.
-          result.newStages.foreach { stage =>
+          // Start materialization of all new stages and fail fast if any stages failed eagerly
+          var anyFail = false
+          val stages = result.newStages.toIterator
+          while (!anyFail && stages.hasNext) {
```

Comment:
because of this shortcut, we can at most have one stage that fails to initialize lazy val?

Reply:
Yeah.

Reply:
And we only need to skip it when canceling stages, right? For other running stages, `stage.resultOption.isEmpty` should be good enough.

Reply:
OK I got it, for other not-started stages, we should not cancel it either.